Sources/StreamChat/Utils/Database/NSManagedObject+Extensions.swift

```diff
@@ -10,13 +10,109 @@ extension NSManagedObject {
+    static func fetchRequest(id: String) -> NSFetchRequest<Self> {
+        fetchRequest(keyPath: idKey, equalTo: id)
+    }
+    
+    static func fetchRequest(keyPath: String, equalTo value: String) -> NSFetchRequest<Self> {
+        let request = NSFetchRequest<Self>(entityName: entityName)
+        request.predicate = NSPredicate(format: "%K == %@", keyPath, value)
+        return request
+    }
+}
+
+extension NSManagedObject: NSFetchRequestGettable {}
+
 extension NSManagedObject {
-    static func load<T>(by request: NSFetchRequest<T>, context: NSManagedObjectContext) -> [T] {
+    static func load<T: NSManagedObject>(by id: String, context: NSManagedObjectContext) -> [T] {
+        load(keyPath: idKey, equalTo: id, context: context)
+    }
+    
+    static func load<T: NSManagedObject>(keyPath: String, equalTo value: String, context: NSManagedObjectContext) -> [T] {
+        let request = NSFetchRequest<T>(entityName: entityName)
+        request.predicate = NSPredicate(format: "%K == %@", keyPath, value)
+        return load(by: request, context: context)
+    }
+    
+    static func load<T: NSManagedObject>(by request: NSFetchRequest<T>, context: NSManagedObjectContext) -> [T] {
+        request.entity = NSEntityDescription.entity(forEntityName: T.entityName, in: context)!
         do {
-            return try context.fetch(request)
+            return try context.fetch(
+                request,
+                using: FetchCache.shared
+            )
         } catch {
             log.error("Could not load \(error), \(error.localizedDescription)")
             return []
         }
     }
 }
+
+class FetchCache {
+    fileprivate static let shared = FetchCache()
+    private let queue = DispatchQueue(label: "io.stream.com.fetch-cache", qos: .userInitiated)
+    private var cache = [NSFetchRequest<NSFetchRequestResult>: [NSManagedObjectID]]()
+
+    fileprivate func set<T>(_ request: NSFetchRequest<T>, objectIds: [NSManagedObjectID]) where T: NSFetchRequestResult {
+        guard let request = request as? NSFetchRequest<NSFetchRequestResult> else {
+            log.assertionFailure("Request should have a generic type conforming to NSFetchRequestResult")
+            return
+        }
+        queue.sync {
+            cache[request] = objectIds
+        }
+    }
+    
+    fileprivate func get<T>(_ request: NSFetchRequest<T>) -> [NSManagedObjectID]? where T: NSFetchRequestResult {
+        guard let request = request as? NSFetchRequest<NSFetchRequestResult> else {
+            log.assertionFailure("Request should have a generic type conforming to NSFetchRequestResult")
+            return nil
+        }
+        var objectIDs: [NSManagedObjectID]?
+        queue.sync {
+            objectIDs = cache[request]
+        }
+        return objectIDs
+    }
+    
+    static func clear() {
+        Self.shared.cache.removeAll()
+    }
+
+    private func clear() {
+        queue.sync {
+            cache.removeAll()
+        }
+    }
```

Comment:
Don't quite understand why we need both? Why isn't `clear()` public? and then we just call `FetchCache.shared.clear()`. Currently, the static method is not thread-safe because it manipulates the cache directly without the `queue`.

Reply:
Oh, that's my mistake. It is supposed to be using the clear instance method! Good catch

Reply:
I fixed the thread-safety